types/grasp/index.d.ts

```diff
@@ -0,0 +1,72 @@
+// Type definitions for grasp 0.6
+// Project: http://graspjs.com
+// Definitions by: Isaac Wolkerstorfer <https://github.com/agnoster>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.1
+
+/// <reference types="node" />
+
+import * as fs from "fs";
+import * as cliColor from "cli-color";
+// Though Grasp does not use esprima, the parser it uses (flow-parser) has a compatible Node type and existing typedefs
+import { Node } from "estree";
+
+export = grasp;
+
+declare function grasp(options: {
+    args: string[] | Record<string, any> | string;
+    error?: (message: string) => void;
+    callback?: (result: string) => void;
+    exit?: (code: number) => void;
+    input?: string;
+
+    // The following are "overrides" for defaults, such as console, cli-color,
+    // process.stdin, or fs. In most cases grasp only makes use of a small
+    // surface area of the types, but I think if someone really wants to pass a
+    // fake fs that only implements the subset of methods they think grasp uses,
+    // it should be up to them to use a cast rather than the typedef to be
+    // overly lenient. We never know if grasp might change their internal use of
+    // fs, after all.
+    fs?: typeof fs;
+    console?: typeof console;
+    textFormat?: typeof cliColor;
```

Comment:
I note Travis is failing on the cli-color dependency. I had considered replacing that with an interface of the methods actually used by the grasp module, but I'm not sure that's a good idea. Would appreciate any thoughts on whether it's better to just have the dependency or use something like

```typescript
textFormat?: Record<"bold" | "magenta" | "red" | "green" | "cyan", (input: string) => string>
```

Reply:
This is suspicious, do they really require this option to implement every method on `cliColor`? Now if `cliColor` added a new method that would be a breaking change. The above line would be preferrable if it only required those five methods.

Reply:
Yeah, for `textFormat`, `fs`, and `console` they only use a small subset of features so it's unclear how to type it. I'm not super comfortable with basing the type on code inspection (grasp could decide to use other methods anytime, basically), so my thinking was that if you really wanted to use those you could use a cast. Basically this whole "give me something that behaves like `fs`" thing feels weird and not ideal for static typing. I'm tempted to just say `any` :-P

Reply:
But yes, it does appear to only use those methods: https://github.com/gkz/grasp/blob/master/src/index.ls#L128

Reply:
Essentially this whole block *could* be:

```typescript
    fs?: Pick<typeof fs, "readFileSync" | "writeFileSync" | "lstatSync" | "readdirSync">;
    console?: Pick<typeof console, "log" | "warn" | "error" | "time" | "timeEnd">;
    textFormat?: Record<"bold" | "magenta" | "red" | "green" | "cyan", (input: string) => string>;
    stdin?: Pick<typeof process.stdin, "setEncoding" | "on" | "resume">;
```

But I'm not sure if that's... better? What do you think? This is my first contribution to DefinitelyTyped so I'm definitely open to suggestions.

Reply:
That's better, although you might be right that giving `fs` a particular type is dangerous if you're not sure exactly how it's supposed to be used.

Reply:
If I do go that route, do you think it would be better to e.g. have an `interface TextFormatter extends Record<...>`, `interface FsLike`, `interface ConsoleLike`, etc. or leave the types inline?

Reply:
Leave them inline.

Reply:
You know the more I think about what it looks like to actually use this the less I'm convinced Pick is a good idea - if you use grasp for "reading a fake file" you'd only really need to implement

```typescript
readFileSync(path: string): string
```

But just to implement `Pick<typeof fs, "readFileSync">` is significantly more complicated.

So I'm leaning back towards leaving as-is with "typeof", and relying on callers who really want to use these features to use e.g. `as typeof fs`. Does that seem okay? Honestly I don't anticipate those options being used much by typescript users, and when they do use them I want them to be aware of the risks.

Reply:
You could use `Partial` -- meaning any methods you do provide must match the signature from `fs`, but you don't have to provide every method.

Reply:
Yeah I considered that, as well as something like `Minimal<T, K extends string> = Pick<T, K> & Partial<T>`, but in the end those still feel just as onerous as `typeof fs` for the reason above (the actual type signatures for most of those methods are significantly more complicated than anyone using them would likely want to do, meaning you'd still need to cast to `typeof fs` even if it were `Partial<typeof fs>`. No one wants to implement

```
readFileSync(path: PathLike | number, options?: { encoding?: null; flag?: string; } | null): Buffer;
readFileSync(path: PathLike | number, options: { encoding: string; flag?: string; } | string): string;
readFileSync(path: PathLike | number, options?: { encoding?: string | null; flag?: string; } | string | null): string | Buffer;
```

when all they need is

```
readFileSync (path: string): string
```

So, at the end of the day, I don't think using `Partial` would really provide a better experience, and would actually potentially be misleading - in that it neither enables you to actually only write the methods you need in the way you need them nor ensures that you didn't miss implementing something that would be needed. So I don't really see the value. I guess I'd either do `typeof fs` or `any` or `{ readFileSync: (path: string) => string, ... }`, and right now I'd lean towards the former.

Does that seem reasonable?

Reply:
:+1: